SETNAME Spec (#361)
ircv3/ircv3-specifications: extensions/setname.md

---
title: IRCv3 `setname` Extension
layout: spec
copyrights:
  -
    name: "Janne Mareike Koschinski"
    period: "2019"
    email: "janne@kuschku.de"
---

## Notes for implementing work-in-progress version

This is a work-in-progress specification.

Software implementing this work-in-progress specification MUST NOT use the

> unprefixed `setname` tag name. Instead, implementations SHOULD use

Reviewer: s/tag name/capability name/, same for the next line

the `draft/setname` capability name to be interoperable with other software
implementing a compatible work-in-progress version.

The final version of the specification will use an unprefixed capability name.

## Motivation

Realnames have historically been immutable. This has never been much of an
issue, as they were rarely used and usually not as present in client UIs.
Still, multiple IRCds have provided independent, non-standard
commands to update realnames.

Nowadays, this has changed. Multiple clients show realnames directly in chat
(e.g. next to the nickname) or use information from the realname to provide
user-visible enhancements (e.g. using emails in the realname to provide
gravatar based avatars).

Due to this, the ability to change the realname has become relevant.

## Description

The `draft/setname` client capability allows clients to change their realname
(GECOS). It also allows servers to directly inform clients about such a change.
This avoids clients having to actually disconnect and reconnect. 

This capability MUST be referred to as `draft/setname` at capability
negotiation time.

> When enabled, clients are allowed to send a `SETNAME` command to notify servers

Reviewer: If the cap isn't enabled, can the client not send a `SETNAME` command? If the client can't use the command, should the command `FAIL` with a more general error code like `CAP_NOT_NEGOTIATED` - message of `"Capability 'draft/setname' is required to use this command, and is not negotiated"` maybe?

Reviewer 2: No, client can always send the command, but won't get a response.

Reviewer: Whichever way is intended, spec language should be clarified to spell it out

Author: Considering the way IRCds implement it currently, I’d suggest that clients can always send it, but they won’t get a response (as unreal/inspircd already implement it).

Reviewer: Slight nibble: The sentence here says _"When enabled, clients are allowed to send a `SETNAME` command..."_. If it's actually able to send the command all the time, should we change this sentence to reflect that? I like the clarification below, makes sense.

---
title: IRCv3 `setname` Extension

> layout: spec

Reviewer 2: Needs the wip tag

```suggestion
layout: spec
work-in-progress: true
```

parameter is the new realname. If the capability is negotiated, the client
MUST assume that no change has happened yet until the server confirms this
change.

Servers MUST check the realname for validity (taking length limits into
consideration). If they accept the realname change, they MUST send the
server-to-client version of the `SETNAME` message to all clients in common
channels, as well as to the client from which it originated.

The `SETNAME` message MUST NOT be sent to clients which do not have the
`draft/setname` capability negotiated.

This server-to-client `SETNAME` message looks as follows:

    :nick!user@host SETNAME :realname goes here

This message represents that the user identified by nick!user@host has changed
their realname to another value. The only, trailing parameter is the new
realname.

In order to take full advantage of the `SETNAME` message, clients must be
modified to support it. The proper way to do so is this:

1) Enable the `draft/setname` capability at capability negotiation time during
   the login handshake.

2) Update the realname portion of data structures and process channel users as
   appropriate.

## Errors

> The server MUST use the FAIL spec to notify the client of failed `SETNAME`

Reviewer 2: Add a link to https://github.com/ircv3/ircv3-specifications/pull/357 and change:

```suggestion
The server MUST use the standard replies extension to notify the client of failed `SETNAME`
```

Reviewer 2: Still needs the link added.

commands.

> If the server rejects the realname due to invalidity, it MUST send a FAIL

Reviewer 2: ```suggestion
If the server rejects the realname as a result of a validation failure, it MUST send a `FAIL`
```

> message of the following format:

Reviewer 2: ```suggestion
message with the `INVALID_REALNAME` code.
```

    FAIL SETNAME INVALID_REALNAME :Realname is not valid
    

> If the server rejects the change for any other reason, it MUST send a FAIL

Reviewer 2: ```suggestion
If the server rejects the change for any other reason, it MUST send a `FAIL`
```

> message indicating this:

Reviewer 2: ```suggestion
message with the `CANNOT_CHANGE_REALNAME` code and an appropriate description of the reason:
```

commands to update realnames.

Nowadays, this has changed. Multiple clients show realnames directly in chat
(e.g. next to the nickname) or use information from the realname to provide
user-visible enhancements (e.g. using emails in the realname to provide
gravatar based avatars).

Due to this, the ability to change the realname has become relevant.

## Description

The `draft/setname` client capability allows clients to change their realname
(GECOS). It also allows servers to directly inform clients about such a change.
This avoids clients having to actually disconnect and reconnect. 

This capability MUST be referred to as `draft/setname` at capability
negotiation time.

Servers advertising the `draft/setname` capability MUST implement the command
as specified here, and MUST support usage of the command even while the
capability is not negotiated.

Servers advertising the capability MUST also provide a `NAMELEN` token in
`RPL_ISUPPORT` with the maximum allowed length for realnames.

If a client sends a `SETNAME` command without having negotiated the capability,
the server SHOULD handle it silently, as the historic implementations did.

> Absence of this capability MUST NOT cause the client to prevent the user from

Reviewer 3: This sentence effectively duplicates the one above:

> and MUST support usage of the command even while the
capability is not negotiated.

Author: What would you then suggest as alternative to replace these paragraphs:

> Servers advertising the `draft/setname` capability MUST implement the command as specified here, and MUST support usage of the command even while the capability is not negotiated.

> If a client sends a `SETNAME` command without having negotiated the capability, the server SHOULD handle it silently, as the historic implementations did.

Reviewer: I'd honestly just leave this paragraph out, both of those paragraphs make different points but do explicitly say that the command is able to be used when the cap isn't negotiated.

Author: One of them says that the command MUST be usable, (but how servers should react to it is still unclear), the other says how they SHOULD react to it. Both are relevant, in a way.

Reviewer: Yepyep. With the first one saying _"MUST support usage of the command even while the capability is not negotiated."_ makes this paragraph redundant. This doesn't add any information, state any new points, or state it in a more understandable way really (that initial line lays it out pretty well).

Reviewer: so the main issue here is just the structure of the sentences was a bit confusing and presented a little backwards (as happens when we get different feedback as we push a spec through the process), made for some misreadings and made it unclear as to why the different sentences were included.

if we re-include text to this effect, a structure along these lines should be less confusing for readers:

>>>
Servers advertising the draft/setname capability MUST support usage of the command even while the capability is not negotiated. Clients MUST NOT prevent users from manually using the command while the capability is not negotiated.

If a client sends a SETNAME command without having negotiated the capability, the server SHOULD handle it silently (with no response), as historic implementations did.

Reviewer 2: Sounds good to me.

layout: spec
copyrights:
  -
    name: "Janne Mareike Koschinski"
    period: "2019"
    email: "janne@kuschku.de"
---

## Notes for implementing work-in-progress version

This is a work-in-progress specification.

Software implementing this work-in-progress specification MUST NOT use the
unprefixed `setname` tag name. Instead, implementations SHOULD use
the `draft/setname` tag name to be interoperable with other software
implementing a compatible work-in-progress version.

The final version of the specification will use an unprefixed tag name.

## Motivation

Realnames have historically been immutable for a connection, which has
historically been accepted, as they were rarely used and usually not as present
in client UIs. Still, multiple IRCds have provided independent, non-standard
commands to update realnames.

Nowadays, multiple clients show realnames directly in chat, next to the nick
name, or use information from the realname to provide enhancements, such as
using emails in the realname to provide gravatar based avatars, leading to
higher demand for this functionality than ever.

## Description

The `draft/setname` client capability allows clients to change their realname

> (GEICOS) as well as allowing servers to directly inform clients about such a

Reviewer: think this should be GECOS rather than GEICOS

Author: Oops, thanks for finding that typo!

change  without having to send a fake quit and join and without clients having
to actually disconnect and reconnect. 

This capability MUST be referred to as `draft/setname` at capability
negotiation time.

> When enabled, clients are allowed to send a `SETNAME` message to notify servers

Reviewer: hmm, chould we refer to it as a 'command' when talking about the client-to-server side? using 'message' is technically correct syntax as per the RFCs, but it feels like saying _"send a `SETNAME` command"_ here (and below when talking about the c2s message) would make it a lil clearer for client authors implementing this

of their intent to update their realname.

This client-to-server `SETNAME` message looks as follows:

    SETNAME :realname goes here

> This message represents the intent to change the realname. The only, trailing

Reviewer: final parameter* rather than trailing. just like with `USER`, the final param doesn't necessarily need to be a trailing one.

Reviewer: alternatively, having an example that very explicitly signifies that the `:` on the final param isn't required would work too, e.g. something like:

    C: SETNAME myrealname
    S: nick!user@host SETNAME myrealname

Reviewer: resolved, thanks!

as specified here, and MUST support usage of the command even while the
capability is not negotiated.

Servers advertising the capability MUST also provide a `NAMELEN` token in
`RPL_ISUPPORT` with the maximum allowed length for realnames. They MUST NOT
apply a different maximum allowed length than the one advertised in

> `RPL_ISUPPORT`.

Reviewer 2: The second sentence sounds a bit redundant, can probably drop it.
